### PR TITLE
Report anonymous roles in authenticate response

### LINKED ISSUE #47195: Authenticate API should return the roles that are inhereted from enabled anonymous access
Authenticated users inherit the roles that are specified for anonymous users. There is already [an issue](https://github.com/elastic/elasticsearch/issues/31589) tracking making this more obvious in the documentation. 

The `_authenticate` API should also return the anonymous roles in the response along with the roles that the authenticated user is assigned and/or mapped to.

### PR DESCRIPTION
**Pull request by ywangd**
Report anonymous roles in response to "GET _security/_authenticate" API call when:
* Anonymous role is enabled
* User is not the anonymous user
* Credentials is not an API Key

Previous attempt (#53453) tried to solve this issue by re-locating anonymous role resolution from authz to authc. It has since been reverted (#57853) because of potential issues on how anonymous access can be used with persisted authentication objects, e.g. Watchers (#57711). Hence the current attempt tries to limit the solution to only the reporting side avoid potential implications. I personally feel this is an acceptable solution and it is also consistent with the original ask.

Resolves: #47195

### CODE REVIEW

**Review by jkakavas — APPROVED** [review]
This looks good to me Yang! Nits are just that, feel free to change or merge as is. Pinging @elastic/kibana-security as they have some tests that expect roles based on the current behavior of the authenticate API and CI will break

File: x-pack/plugin/security/qa/basic-enable-security/src/test/java/org/elasticsearch/xpack/security/EnableSecurityOnBasicLicenseIT.java [hunk 1]
```diff
@@ -132,7 +132,7 @@ private void checkAuthentication() throws IOException {
         final Map<String, Object> auth = getAsMap("/_security/_authenticate");
         // From file realm, configured in build.gradle
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/qa/basic-enable-security/src/test/java/org/elasticsearch/xpack/security/EnableSecurityOnBasicLicenseIT.java:133]: Add a comment about anonymous access is configured in build.gradle

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/user/TransportAuthenticateActionTests.java [hunk 2]
```diff
@@ -118,7 +122,33 @@ public void onFailure(Exception e) {
...
     }
+
+    private AnonymousUser prepareAnonymousUser() {
+        final AnonymousUser anonymousUser = mock(AnonymousUser.class);
+        if (randomBoolean()) {
+            when(anonymousUser.enabled()).thenReturn(true);
+        } else {
+            when(anonymousUser.enabled()).thenReturn(false);
+        }
+        final String[] roleNames = randomList(1, 4, () -> randomAlphaOfLengthBetween(4, 12)).toArray(new String[0]);
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/user/TransportAuthenticateActionTests.java:149]: You right. Thanks

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportAuthenticateAction.java [hunk 3]
```diff
@@ -43,7 +49,32 @@ protected void doExecute(Task task, AuthenticateRequest request, ActionListener<
...
         } else {
-            listener.onResponse(new AuthenticateResponse(authentication));
+            final User user = authentication.getUser();
+            final boolean shouldAddAnonymousRoleNames = anonymousUser.enabled() && false == anonymousUser.equals(user)
+                && authentication.getAuthenticationType() != Authentication.AuthenticationType.API_KEY;
+            if (shouldAddAnonymousRoleNames) {
+                final String[] allRoleNames = Stream.concat(
+                    Stream.of(user.roles()), Stream.of(anonymousUser.roles())).toArray(String[]::new);
+                listener.onResponse(new AuthenticateResponse(
+                    new Authentication(
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportAuthenticateAction.java:59]: I initially had this logic inside the `User` class, i.e. `User#withRoles` with the intention to simply things here. But as you said, there is no generic need for this type of method, which made it seem out of place. So I decided to keep everything in this one class. It is verbose but at least self-contained and hopefully easy to understand.

**Pull request merged**